pkg/hive/hive_test.go

```diff
@@ -0,0 +1,170 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package hive
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"testing"
+
+	ma "github.com/multiformats/go-multiaddr"
+
+	"github.com/ethersphere/bee/pkg/addressbook/inmem"
+	"github.com/ethersphere/bee/pkg/discovery/mock"
+	pb "github.com/ethersphere/bee/pkg/hive/pb"
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/p2p"
+	"github.com/ethersphere/bee/pkg/p2p/protobuf"
+	"github.com/ethersphere/bee/pkg/p2p/streamtest"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+func TestInit(t *testing.T) {
+	logger := logging.New(ioutil.Discard, 0)
+	connecter := &ConnecterMock{}
+	peerer := mock.NewPeerer()
+	addressBook := inmem.New()
+
+	// this is the receiving side
+	nodeReceiver := New(Options{
+		Peerer:      peerer,
+		AddressBook: addressBook,
+		Logger:      logger,
+	})
+
+	// setup the stream recorder to record stream data
+	streamer := streamtest.New(
+		streamtest.WithProtocols(nodeReceiver.Protocol()),
+	)
+
+	nodeInit := New(Options{
+		Streamer:  streamer,
+		Connecter: connecter,
+		Logger:    logger,
+	})
+
+	overlays := []swarm.Address{swarm.MustParseHexAddress("1aaaaaaa"),
+		swarm.MustParseHexAddress("1bbbbbbb"),
+		swarm.MustParseHexAddress("1ccccccc")}
+	underlays := []ma.Multiaddr{newMultiAddr("/ip4/1.1.1.1"),
+		newMultiAddr("/ip4/1.1.1.2"),
+		newMultiAddr("/ip4/1.1.1.3")}
+	testPeer := p2p.Peer{Address: swarm.MustParseHexAddress("ca1e9f3a")}
+
+	//populate discovery peerer for bin 1 & 2
+	peerer.Add(testPeer, 0, p2p.Peer{Address: overlays[0]}, p2p.Peer{Address: overlays[1]})
+	peerer.Add(testPeer, 1, p2p.Peer{Address: overlays[2]})
+
+	// populate address book
+	for i := 0; i < len(overlays); i++ {
+		addressBook.Put(overlays[i], underlays[i])
+	}
+
+	t.Run("OK - node responds with some peers for bin 1 & 2", func(t *testing.T) {
+		if err := nodeInit.Init(context.Background(), testPeer); err != nil {
+			t.Fatal(err)
+		}
+
+		records, err := streamer.Records(testPeer.Address, protocolName, protocolVersion, peersStreamName)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if l := len(records); l != maxPO {
+			t.Fatalf("got %v records, want %v", l, maxPO)
+		}
+
+		// validate received requestPeers requests
+		// there should be maxPO peer requests
+		var wantGetPeers []*pb.GetPeers
+		var gotGetPeers []*pb.GetPeers
+		for i := 0; i < maxPO; i++ {
+			wantGetPeers = append(wantGetPeers, &pb.GetPeers{
+				Bin:   uint32(i),
+				Limit: 10,
+			})
+
+			messages, err := readAndAssertMessages(records[i].In(), 1, func() protobuf.Message {
+				return new(pb.GetPeers)
+			})
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			gotGetPeers = append(gotGetPeers, messages[0].(*pb.GetPeers))
+		}
+
+		if fmt.Sprint(gotGetPeers) != fmt.Sprint(wantGetPeers) {
+			t.Errorf("requestPeers got %v, want %v", gotGetPeers, wantGetPeers)
+		}
+
+		// validate Peers response
```

Comment:
this could be the table driven test. 

since you're testing multiple different scenarios (i.e. bin 0 should yield this and that, bin 1 should not, bin 16 should not etc) - you could just have a table driven test that checks isolated bins and asserts the response that the protocol should yield in any given case.

this format of test is also problematic because in line `L110` you're _relying_ on the output of the actual mock in order to assert the reply. That's all good and well, but if you change the implementation in the mock, or actually everywhere else you will not be able to detect any regressions with this test as it will just verify that the mock works according to its own logic, otherwise nothing else is being tested here.

Reply:
Well, this is actually a single scenario where I expect bin 0 and bin 1 to be populated and others not, because that is how `Init` function works here. Different scenarios would be different maps for different combinations of responses (different bins populated, etc...), and expecting different results. In that case, I agree, I would use table driven tests, and if I go in that direction, I would transform this into table driven tests that would look similar as this, but with different maps as inputs and different results in assertions driven by tables.
Regarding the mock, I actually think that if anything is changed in mock or anywhere, the test would fail, which is exactly what I wanted. Then, the code in the tests would need to be changed as well. Could you explain a bit more what you feel is problematic here?

Reply:
I'm not sure you got what I meant.
You populate the mock in line 59-60, but in the test below in line 110 you make no hard assertions about what comes out of the mock (if the mock returns nothing for bin 0 or bin 1 the test will not break). it is OK for the higher bins since the loop that starts at line 112 will no run (since there would be no elements in `p`), but it does not guarantee that the test will break on bin 0 and 1.
Personally I find using a declared expectation on what's supposed to be found in any bin. Using `for range` is nice and clean but often produces situations where the loop does not run resulting in some unexpected results and false positives in test results

Reply:
I see what you mean, good point, the assertion should be more precise! Nice catch. I will close this PR anywayz in favor of the alternative hive one, but I will keep this in mind for the future, thanks!